ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/TypeHintsAdornmentProvider.fs

```diff
@@ -0,0 +1,125 @@
+open JetBrains.UI.RichText
+open FSharp.Compiler.SourceCodeServices
+open FSharp.Compiler.Layout
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Tree
+open JetBrains.ReSharper.Psi
+open JetBrains.ReSharper.Psi.Modules
+
+[<DaemonIntraTextAdornmentProvider(typeof<TypeHintsAdornmentProvider>)>]
+[<StaticSeverityHighlighting(Severity.INFO,
+     HighlightingGroupIds.IntraTextAdornmentsGroup,
+     AttributeId = AnalysisHighlightingAttributeIds.PARAMETER_NAME_HINT,
+     OverlapResolve = OverlapResolveKind.NONE,
+     ShowToolTipInStatusBar = false)>]
+type TypeHintHighlighting(text: RichText, range: DocumentRange) =
+    interface IHighlighting with
+        member x.ToolTip = null
+        member x.ErrorStripeToolTip = null
+        member x.IsValid() = not text.IsEmpty && not range.IsEmpty
+        member x.CalculateRange() = range
+
+    interface IHighlightingWithTestOutput with
+        member x.TestOutput = text.ToString()
+
+    member x.Text = text
+
+and [<SolutionComponent>] TypeHintsAdornmentProvider() =
+    interface IHighlighterIntraTextAdornmentProvider with
+        member x.CreateDataModel(highlighter) =
+            match highlighter.UserData with
+            | :? TypeHintHighlighting as thh ->
+                { new IIntraTextAdornmentDataModel with
+                    override x.Text = thh.Text
+                    override x.HasContextMenu = false
+                    override x.ContextMenuTitle = null
+                    override x.ContextMenuItems = null
+                    override x.IsNavigable = false
+                    override x.ExecuteNavigation _ = ()
+                    override x.SelectionRange = Nullable<_>()
+                    override x.IconId = null
+                    override x.IsPreceding = false
+                }
+            | _ -> null
+
+type TypeHighlightingVisitor(fsFile: IFSharpFile, checkResults: FSharpCheckFileResults) =
+    inherit TreeNodeVisitor<IHighlightingConsumer>()
+
+    let tokenNames = ["|>"]
+
+    override x.VisitNode(node, context) =
+        for child in node.Children() do
+            match child with
+            | :? IFSharpTreeNode as treeNode -> treeNode.Accept(x, context)
+            | _ -> ()
+
+    override x.VisitBinaryAppExpr(binding, consumer) =
+        let opExpr = binding.Operator
+        if opExpr.QualifiedName <> "|>" then () else
+
+        match opExpr.Identifier.As<FSharpIdentifierToken>() with
+        | null -> ()
+        | token ->
+
+            let sourceFile = opExpr.GetSourceFile()
+            let coords = sourceFile.Document.GetCoordsByOffset(opExpr.GetTreeEndOffset().Offset)
+            let lineText = sourceFile.Document.GetLineText(coords.Line)
+            use cookie = CompilationContextCookie.GetOrCreate(fsFile.GetPsiModule().GetContextFromModule())
+
+            let getTooltip = checkResults.GetStructuredToolTipText(int coords.Line + 1, int coords.Column, lineText, tokenNames, FSharpTokenTag.Identifier)
```

Comment:
It would be great to have this logic extracted as a static method (e.g. taking a token and other needed things as params) in `FSharpIdentifierTooltipProvider` or in a similar place.